### ISSUE TITLE
Fix S1854 FP: Recursive functions

### ISSUE THREAD

**Issue by hankovich** [SonarSource/sonar-dotnet#7691]
### Description

S1854 reportes false positive for recursive functions

### Repro steps

Case 1

```csharp
var visited1 = false;

void Visit1()
{
    if (visited1)
    {
        return;
    }

    visited1 = true;

    var a = float.Epsilon; // <- S1854 remove the useless assignment to local variable 'a'

    if (Random.Shared.Next() > 0)
    {
        Visit1();

        Console.WriteLine(a);

        return;
    }

    Visit1();

    Console.WriteLine(a);
}
```

Case 2

```csharp
var visited2 = false;

void Visit2()
{
    if (visited2)
    {
        return;
    }

    visited2 = true;

    var a = false; // no diagnostic here

    if (Random.Shared.Next() > 0)
    {
        a = true; // <- S1854 remove the useless assignment to local variable 'a'
    }

    Visit2();

    Console.WriteLine(a);
}
```

It looks like booleans are processed in a different way comparing to other types, because if I change `float` to `bool` in case 1, then no diagnostic is reported.

### Expected behavior

No S1854 diagnotic is reported.

### Actual behavior

S1854 diagnotic is reported.

### Known workarounds

N/A.

### Related information

* C#/VB.NET Plugins version 9.5.0.73987
* Visual Studio version latest preview
* MSBuild / dotnet version .NET 8 preview 6
* SonarScanner for .NET version (if used) N/A
* Operating System Windows 10

**Comment by cristian-ambrosini-sonarsource** [comment]
Hi @hankovich, the first example you provided is not raising S1854 (`a` is passed as a parameter to the `Console.WriteLine` method so it's correctly not reporting an issue). But I found indeed a FN while playing with a couple of reproducers:

```cs
public class Repro_7691
{
    private bool flag = false;

    void MyMethod()
    {
        var a = false; // Compliant
        var b = float.Epsilon; // Compliant

        if (flag)
        {
            a = true; // Compliant
            b = float.MaxValue; // Compliant
        }

        Console.WriteLine(a);
        Console.WriteLine(b);
    }
}
```

```cs
public class Repro_7691_2
{
    private bool flag = false;

    void MyMethod()
    {
        var a = false; // Compliant FN <----------------------------
        var b = float.Epsilon; // Noncompliant

        a = true; // Noncompliant
        b = float.MaxValue; // Noncompliant
    }
}
```

If it's ok with you I'll rename this issue to target this FN and add a reproducer.

**Comment by hankovich** [comment]
Hi. I'm happy to hear that I helped to detect a couple of FNs, but the original issue was about FPs which I have in my projects. I'll try to provide a new repro.

**Comment by cristian-ambrosini-sonarsource** [comment]
I'm closing this issue for now due to inactivity. @hankovich, if you manage to find a new reproducer, please feel free to open a new issue. Thank you!
